tensorflow/compiler/xla/service/gpu/ir_emitter_unnested.cc

```diff
@@ -1978,32 +1991,68 @@ void IrEmitterUnnested::EmitTile(
-                   source_idx.AddOffsetToDim(y_loc, kDimY, &b_)
-                       .AddOffsetToDim(constant(j * step_x), kDimX, &b_);
-               auto emit_element = [&] {
-                 return emit_elem_function(source_idx_x, y_loc, x_loc, j);
-               };
-               if (!x_tile_fits) {
-                 ksl->If(loop_name + "_x_in_tile",
-                         b_.CreateICmpULT(x_loc, tile_width), emit_element);
-               } else {
-                 emit_element();
-               }
-             }
-           });
+  int vector_size = mapping_scheme.GetVectorSize();
+  ksl->For(
+      loop_name + "_y_in_tile",
+      /*start=*/constant(0),
+      /*end=*/
+      ceil_of_ratio(b_.CreateSub(tile_height, thread_id_info.thread_id_y),
+                    num_threads_y),
+      /*step=*/constant(1), [&](llvm::Value* y_indvar) {
+        llvm::Value* y_loc = b_.CreateAdd(
+            thread_id_info.thread_id_y, b_.CreateMul(y_indvar, num_threads_y));
+        auto unroll = [&](bool add_if, int64 max_element, int64 vector_size) {
+          for (int64 j = 0; j < x_num_steps / vector_size; j++) {
+            // Prep some values. If we do not do this, LLVM doesn't vectorize.
+            llvm::Value* x_loc_base =
+                b_.CreateAdd(constant(j * step_x * vector_size), start_offset_x,
+                             "x_loc_base");
+            IrArray::Index source_idx_x_base =
+                source_idx.AddOffsetToDim(y_loc, kDimY, &b_)
+                    .AddOffsetToDim(constant(j * step_x * vector_size), kDimX,
+                                    &b_);
+
+            for (int i = 0; i < vector_size; i++) {
+              int old_j = j * vector_size + i;
+              llvm::Value* x_loc = x_loc_base;
+              IrArray::Index source_idx_x = source_idx_x_base;
+              if (i > 0) {
+                x_loc = b_.CreateAdd(constant(i), x_loc_base, "x_loc");
+                source_idx_x =
+                    source_idx_x_base.AddOffsetToDim(constant(i), kDimX, &b_);
+              }
+              auto emit_element = [&] {
+                return emit_elem_function(source_idx_x, y_loc, x_loc, old_j);
+              };
+              if (add_if) {
+                ksl->If(loop_name + "_x_in_tile",
+                        b_.CreateICmpULT(x_loc, tile_width), emit_element);
+              } else {
+                emit_element();
+              }
+            }
+          }
+        };
+
+        if (!x_tile_fits && mapping_scheme.GetIndexingOrder() ==
+                                KernelMappingScheme::LinearDilatedIndexingX) {
+          // Only try this path when we try to vectorize the loads.
+
+          // Special case when the tile doesn't fit completly for even row size.
+          // For odd row size every other row isn't aligned, so can't be
+          // vectorized.
+          ksl->If(loop_name + "_is_full_tile",
+                  // if (block fully fit) {fast path} else {slow path}
```

Comment:
Do we actually need this fast_path/slow_path divergence? I understand the optimization, but previously I have removed a lot of those, because they add a lot of code complexity, but often lead to negligible increases in performance.

Reply:
This is needed. See my answer to comment: https://github.com/tensorflow/tensorflow/pull/37260#discussion_r387805494

Reply:
Sorry I'm still a bit confused there, are you saying that when `add_index_boundary_conditions` is `false`, vectorization does not happen? Do we even need unrolling in that case then?

Reply:
For LLVM to vectorize even row size, we need 2 things to have the vectorized loads and handle the partial tile at the end of the row.
- One path without the condition between each loads so that LLVM vectorize them.
- One path with the condition to manage for the last warp that loads only part of the elements.

The add_index_boundary_conditions when false, won't add the condition between loads. So LLVM can vectorize the loads. But when true, it handle the partial tile at the end of the row.